src/importer.rs

```diff
@@ -0,0 +1,284 @@
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::fs;
+use std::io::ErrorKind;
+
+use crc32fast;
+use tobj;
+
+use crate::viewport_renderer::{Index, Vertex};
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ImporterError {
+    FileNotFound,
+    PermissionDenied,
+    InvalidStructure,
+    Other,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Model {
+    pub name: String,
+    pub vertices: Vec<Vertex>,
+    pub indices: Vec<Index>,
+}
+
+/// `Importer` takes care of importing of obj files and caching of their
+/// internal representations. It holds paths to files, their checksums and
+/// parsed obj files.
+#[derive(Debug, Default)]
+pub struct Importer {
+    pub path_checksums: HashMap<String, u32>,
+    pub checksum_paths: HashMap<u32, Vec<String>>,
+    pub loaded_models: HashMap<String, Vec<Model>>,
+}
+
+impl Importer {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Tries to import obj file from given `path`. If file was already
+    /// imported and its checksum is identical, parsed models are returned
+    /// from cache. If the checksum doesn't match, file is reimported and
+    /// the old one is forgotten. If file wasn't imported, but its checksum
+    /// matches other file, it is cloned. Brand new files are parsed and
+    /// cached.
+    pub fn import_obj(&mut self, path: &str) -> Result<Vec<Model>, ImporterError> {
```

Comment:
This could return a `Cow<Vec<Model>>` in the Ok variant if you want to avoid further clones, but let's not go there just yet 😉